lib/clarivate/links_client.rb

```diff
@@ -0,0 +1,65 @@
+module Clarivate
+  # Links AMR (Article Match Retrieval) Service
+  # @see http://ipscience-help.thomsonreuters.com/LAMRService/WebServicesOverviewGroup/overview.html Service documentation
+  class LinksClient
+    attr_reader :username, :password, :host
+
+    def initialize(username: nil, password: nil, host: 'https://ws.isiknowledge.com')
+      @host = host
+      @username = username
+      @password = password
+      (@username, @password) = Base64.decode64(Settings.WOS.AUTH_CODE).split(':', 2) unless username
+    end
+
+    # limited to 50 ids
+    # @param [Array<String>] ids
+    # @return [Hash<String => Hash>]
+    def links(ids)
+      raise ArgumentError, "1-50 ids required" unless ids.present? && ids.count <= 50
+      response = connection.send(:post) do |req|
+        req.path = '/cps/xrpc'
+        req.body = request_body(ids.uniq)
+      end
+      ng = Nokogiri::XML(response.body) { |config| config.strict.noblanks }.remove_namespaces!
+      pairs = ng.xpath('response/fn/map/map').map do |node|
+        [node.attr('name').split('_').last, vals_to_hash(node.xpath('//val'))]
+      end
+      pairs.to_h
+    end
+
+    private
+
+      # @return [Faraday::Connection]
+      def connection
+        @connection ||= Faraday.new(url: host) do |faraday|
+          faraday.use Faraday::Response::RaiseError
+          faraday.adapter :httpclient
+        end
+      end
+
+      # @return [ActionView::Base] used to render as though in an rails controller
+      def renderer
+        @renderer ||= ActionView::Base.new(ActionController::Base.view_paths, {})
+      end
+
+      def request_body(ids)
+        renderer.render(
+          file: 'pages/clarivate_links.xml',
+          layout: false,
+          locals: {
+            client: self,
+            return_fields: %w(ut doi pmid),
+            cites: ids
+          }
+        )
+      end
+
+      # @param [<Nokogiri::XML::Nodeset>] vals
+      # @return [Hash<Symbol => String>] namespace to value
+      # Nodeset for [<val name="ut">UT value</val>, <val name="doi">DOI</val>]
+      # becomes { ut: 'UT value', doi: 'DOI'}
+      def vals_to_hash(vals)
+        vals.map { |val| [val.attr("name").downcase.to_sym, val.text.strip] }.to_h
+      end
```

Comment:
Why all the massage for `val.attr("name").downcase.to_sym`?  Simply `val.attr("name")` should be OK, so the client doesn't mask anything in the data.  Do we need the `ut` in the vals - can we assume that their response map is always reliable?  Can we assume that there can only be one `ut` and if we have it in the hash-key, that's enough?

Reply:
Because I wanted the Hash to be keyed by symbols (`val.attr("name")` is a string), and I wanted to be strict about what goes in.

Reply:
Do we know whether anything could be lost by a `downcase` and whether there is any inconsistency in the API key strings?  (I'm considering millions of these strings going through a `downcase` that may not be necessary.)

Reply:
`# @param [<Nokogiri::XML::Nodeset>] vals` - probably works, I dunno; maybe the `<>` are not necessary, i.e.
`# @param [Nokogiri::XML::Nodeset] vals` is it (looks like a copy-n-paste of a .class).

Reply:
Ah, good catch.  Leftover from when I was first taking an `Array<...>`.